# learning bioinformation-  3.24 differencial analysis
 - 学习内容来自[KeepNotes blog](https://www.bioinfo-scrounger.com/archives/499/)
 
## Begain
  
  `install.packages("rpx")`

 - 查看[proteomexchange](http://www.proteomexchange.org/)新发布的数据:
`pxannounced()`

- 对质谱数据进行分析比对（确定多肽段的蛋白&基因名称）即为查库

 - 查库软件：
#对于raw data————try `MSnbase`

`install.packages("MSnbase)`

## 统计分析：R包*3
### step1 安装包
`install.packages("MSstats")`
 > for statistical relative protein signigicance analysis in **DDA,SRM and DIA.data** stored in **data.fram or MSnSet** can be used as input 
 
 `install.packages("msmsTests")`


> Statistical tests for **label-free** LC-MS/MS data by 
  **spectral counts**, to discover differentially expressed proteins between two biological conditions. Three tests are available: Poisson GLM regression, quasi-likelihood GLM regression, and the negative binomial of the edgeR package. All can be readily applied on MSnSet instances produced, for example by MSnID  
  
`install.packages("isobar")`

> This one also provides dedicated infrastructure for the statistical analysis of **isobaric data**

## step2 参考 [KeepNotes blog](https://www.bioinfo-scrounger.com/archives/541/),尝试进行差异分析（网站下载数据文件）

> library(ggplot2)       
setwd("D:/test_blog_differencially")  
test_data <- read.csv(file = "labelfree.csv", sep = ",", header = T, row.names = 1)

 - Q: load csv多了一列X全部为NA值？
 - A: 数据源有问题，打开excel后选中最后几列，删除后再导入，成功

> rm <- apply(test_data, 1, function(x){sum(x == 0) > 3
})  
df <- test_data[!rm,]  
nrow(df)
 - nrow(df)这里开始出现问题，和原文结果有差异，果然，后面出岔子了

 - 为什么进行log转化？ 
 - 丰度值数据分布非常离散,可尝试计算均值和方差
 
> Mean <- apply(df, 2, mean)  
SD <- apply(df, 2, sd)  
Mean  
SD
 - 差异分析使用何种数据？
  - >limma & MSstats必须进行对数转化，前者默认输入数据已经做了log转化，后者根据参数来设定如何进行log转化  
 另：**T检验或者方差分析**需要数据符合**方差齐性**，实验过程中误差可能越来越大，因此对数变化可以将方差变大的趋势调回来

### 尝试进行T test
 - 需要样本总体符合正态分布，并且为方差齐性，首先进行F检验
> fvalue <- apply(df, 1, function(x){  
      a <- factor(c(rep("A", 3), rep("B", 3)))  
  var.test(x~a)
})  
sum(unlist(lapply(fvalue, function(x){
  x$p.value > 0.05
})))

 - 结果表示5325个蛋白中有4605个满足方差齐性

 - t test进行**成组T检验，双尾检验**。若为方差不齐，则用`var.equal = FALSE`

`pvalue <- apply(df, 1, function(x){  
a <- factor(c(rep("A",3), rep("B",3)))  
   fvalue <-var.test(x~a)  
  if (fvalue$p.value >0.05){  
    t.test(x~a, var.equal = T)  
  }else{   
    t.test(x~a, var.equal = F)  
  }
})`

result_ttest <- data.frame(ID=row.names(df),
                           Pvalue = as.numeric(unlist(lapply(pvalue,function(x) x$p.value))),
                           log2FC = log2(as.numeric(unlist(lapply(pvalue, function(x) x$estimate[1]/x$estimate[2]))))
                           )

sum(result_ttest$Pvalue < 0.05 & abs(result_ttest$log2FC) > 1)





## 1 下载安装biocontruct
 - Install/Update sandpaper

>options(repos = c(carpentries = "https://carpentries.r-universe.dev/", 
                  CRAN = "https://cloud.r-project.org"))

> install.packages("sandpaper")



 - update the workflows 
> library("sandpaper")
update_github_workflows()

 - update my R and Rstudio

> install.packages('installr')
library('installr')

> updateR()

报错: cannot find “string”

> install.packages("stringi")
library(stringi)

 - install bioconduct
> if (!require("BiocManager", quietly = TRUE))
  
  >install.packages("BiocManager")
BiocManager::install(version = "3.16")
 - 测试bioconduct安装情况
> BiocManager::install(c("GenomicFeatures", "AnnotationDbi"))
BiocManager::available()


## 2 MS analysis
 - install MSstats
 > if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("MSstats")

> browseVignettes("MSstats")



 - Attention! 
 
 'MSstatsInput.csv' is the MSstats report from [Skyline](https://skyline.ms/project/home/software/Skyline/begin.view). 
 
 PS: skyline can analysis .raw data and output .csv 

> input <- read.csv(file="MSstatsInput.csv")

> raw <- SkylinetoMSstatsFormat(input)

 -  MSstats can also Read in MaxQuant files

eg:
> proteinGroups <- read.table("proteinGroups.txt", sep     ="\t", header=TRUE)

>  infile <- read.table("evidence.txt", sep="\t", header=TRUE)

>/#  Read in annotation including condition and biological replicates per run.                                       
>/# Users should make this annotation file. It is not the output from MaxQuant.  ??? **don't understand**

>annot <- read.csv("annotation.csv", header=TRUE)
   
> raw <- MaxQtoMSstatsFormat(evidence=infile,                          annotation=annot,
 proteinGroups=proteinGroups)
 
 - practice MSstat>>dataProcess with .raw data
> library("MSstats")  
dataProcess()


**error**:The following object is masked from ‘package:grDevices’:savePlot

### 2.1 try dataProcess

>dataProcess(
  "2003151_WHJ_CSF_TMT_11.raw", logTrans = 2,  
  normalization = "equalizeMedians",  
  nameStandards = NULL,  
  featureSubset = "all",  
  remove_uninformative_feature_outlier = FALSE,  
  min_feature_count = 2,  
  n_top_feature = 3,  
  summaryMethod = "TMP",  
  equalFeatureVar = TRUE,  
  censoredInt = "NA",  
  MBimpute = TRUE,  
  remove50missing = FALSE,  
  fix_missing = NULL,  
  maxQuantileforCensored = 0.999,  
  use_log_file = TRUE,  
  append = FALSE,  
  verbose = TRUE,  
  log_file_path = NULL  
  )

 - error: 
 >Error in (function (classes, fdef, mtable)  :   
函数‘.checkDataValidity’标签‘"character"’找不到继承方法
 - reason: 函数无法识别输入参数的类型，即 dataprocess无法识别.raw data

 - **try .txt**
 
> getwd()  
setwd("D:/test_example/combined/txt")    
proteinGroups <- read.table("proteinGroups.txt", sep="\t", header=TRUE)

>dataProcess("proteinGroups.txt",logTrans = 2)

 - error:函数‘.checkDataValidity’标签‘"character"’找不到继承方法



###Mfuzz
BiocManager::install("Mfuzz")
browseVignettes("Mfuzz")
n

# 3.24 继续bioconduct
 - 聚类分析直接用cluster包
> library("cluster")

> library("ggplot2")

update package
>old.packages()  
> update.packages()


















































